### PR TITLE
merge bitcoin#15588...#16475: backports

### PR DESCRIPTION
**Pull request by kittywhiskers**

### CODE REVIEW

**Comment by PastaPastaPasta** [comment]
build fails

**Review by PastaPastaPasta — CHANGES_REQUESTED** [review]
see prior

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

**Comment by kittywhiskers** [comment]
## Current Status 

* `wallet_balance.py`'s behavior is non-deterministic and half the code is commented out
* `wallet_resendwallettransactions.py` just straight up fails

File: test/functional/wallet_basic.py [hunk 1]
```diff
@@ -222,24 +222,10 @@ def run_test(self):
...
-        # EXPECT: nodes[3] should have those transactions in its mempool.
-        txid1 = self.nodes[0].sendtoaddress(self.nodes[1].getnewaddress(), 1)
-        txid2 = self.nodes[1].sendtoaddress(self.nodes[0].getnewaddress(), 1)
-        self.sync_mempools(self.nodes[0:2])
-
         self.start_node(3)
         connect_nodes_bi(self.nodes, 0, 3)
         self.sync_blocks()
 
-        relayed = self.nodes[0].resendwallettransactions()
```
[the remark this answers is not in the document]
Comment (kittywhiskers) [reply test/functional/wallet_basic.py:237]: This PR is already pretty large, I can backport the commits that do the remaining `sync_block` changes if we split the PR (we're at ~16 commits already)

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
- 15620 is still missing some changes (requires 15401)
- drop 6097917 and bring back 15646 + https://github.com/UdjinM6/dash/commit/fbce51ca7d6fda2005bd87ad4c046a875bc5fd30
- drop 16961 (was backported earlier)
- see https://github.com/UdjinM6/dash/commit/abe35f6139deb68e5908bf697e1423b464f16113

File: test/functional/wallet_resendwallettransactions.py [hunk 2]
```diff
@@ -2,31 +2,77 @@
...
 
-        # Should return an array with the unconfirmed wallet transaction.
-        txid = self.nodes[0].sendtoaddress(self.nodes[0].getnewaddress(), 1)
-        assert_equal(self.nodes[0].resendwallettransactions(), [txid])
+        self.log.info("Transaction should be rebroadcast after 30 minutes")
+        # Use mocktime and give an extra 5 minutes to be sure.
+        rebroadcast_time = self.mocktime + 41 * 60
+        node.setmocktime(rebroadcast_time)
+        self.mocktime = rebroadcast_time
+        def wait_p2p_1():
```
Comment (UdjinM6) [inline test/functional/wallet_resendwallettransactions.py:78]: 15646: I messed up in my fix a bit it seems 🙈 
```
test/functional/wallet_resendwallettransactions.py:78:9: E306 expected 1 blank line before a nested definition, found 0
^---- failure generated from test/lint/lint-python.sh
```
```suggestion

        def wait_p2p_1():
```

File: src/rpc/rawtransaction.cpp [hunk 3]
```diff
@@ -809,14 +809,15 @@ UniValue sendrawtransaction(const JSONRPCRequest& request)
 
     bool bypass_limits = false;
     if (!request.params[3].isNull()) bypass_limits = request.params[3].get_bool();
-    uint256 txid;
     std::string err_string;
-    const TransactionError err = BroadcastTransaction(tx, txid, err_string, max_raw_tx_fee, bypass_limits);
+    AssertLockNotHeld(cs_main);
+
```
Comment (PastaPastaPasta) [inline src/rpc/rawtransaction.cpp:814]: nit: 15713: there shouldn't be a newline here

File: src/Makefile.bench.include [hunk 4]
```diff
@@ -83,7 +85,7 @@ CLEAN_BITCOIN_BENCH = bench/*.gcda bench/*.gcno $(GENERATED_BENCH_FILES)
 
 CLEANFILES += $(CLEAN_BITCOIN_BENCH)
 
-bench/checkblock.cpp: bench/data/block813851.raw.h
+bench/data.cpp: bench/data/block813851.raw.h
```
Comment (PastaPastaPasta) [inline src/Makefile.bench.include:88]: 16299: missing change here https://github.com/bitcoin/bitcoin/pull/16299/files#diff-42a43de260ebffc4d0cc72815d4dc4780b26af35c32266c4eaa45196e6b70fdaR94

File: src/interfaces/wallet.cpp [hunk 5]
```diff
@@ -317,12 +291,28 @@ class WalletImpl : public Wallet
     {
         auto locked_chain = m_wallet->chain().lock();
         LOCK2(mempool.cs, m_wallet->cs_wallet);
-        auto pending = MakeUnique<PendingWalletTxImpl>(*m_wallet);
-        if (!m_wallet->CreateTransaction(*locked_chain, recipients, pending->m_tx, pending->m_key, fee, change_pos,
+        CReserveKey m_key(m_wallet.get());
```
Comment (PastaPastaPasta) [inline src/interfaces/wallet.cpp:294]: I think this is fine, just noting difference here

File: src/qt/walletmodeltransaction.cpp [hunk 6]
```diff
@@ -22,14 +22,14 @@ QList<SendCoinsRecipient> WalletModelTransaction::getRecipients() const
...
-std::unique_ptr<interfaces::PendingWalletTx>& WalletModelTransaction::getWtx()
+CTransactionRef& WalletModelTransaction::getWtx()
 {
     return wtx;
 }
 
 unsigned int WalletModelTransaction::getTransactionSize()
 {
-    return wtx ? ::GetSerializeSize(wtx->get(), SER_NETWORK, PROTOCOL_VERSION) : 0;
+    return wtx != nullptr ? ::GetSerializeSize(wtx.get(), SER_NETWORK, PROTOCOL_VERSION) : 0;
```
Comment (PastaPastaPasta) [inline src/qt/walletmodeltransaction.cpp:32]: adding the `!= nullptr` is okay, maybe even desired, but is also kinda unrelated here

Comment (PastaPastaPasta) [reply src/qt/walletmodeltransaction.cpp:32]: 16415: also
```suggestion
    return wtx != nullptr ? ::GetSerializeSize(*wtx, SER_NETWORK, PROTOCOL_VERSION) : 0;
```

**Review by UdjinM6 — DISMISSED** [review]
LGTM, utACK

**Review by PastaPastaPasta — CHANGES_REQUESTED** [review]

File: src/qt/walletmodeltransaction.cpp (hunk 6 above)
Comment (kittywhiskers) [reply src/qt/walletmodeltransaction.cpp:32]: AFAIK, the lack of `!= nullptr` causes a build failure and was therefore added

**Review by PastaPastaPasta — APPROVED** [review]
utACK for merging via merge commit

**Review by UdjinM6 — APPROVED** [review]
re-utACK

**Pull request merged**